src/librustc_trans/trans/_match.rs

```diff
@@ -329,11 +329,35 @@ pub enum OptResult<'blk, 'tcx: 'blk> {
 
 #[derive(Clone, Copy, PartialEq)]
 pub enum TransBindingMode {
+    /// By-value binding for a copy type: copies from matched data
+    /// into a fresh LLVM alloca.
     TrByCopy(/* llbinding */ ValueRef),
-    TrByMove,
+
+    /// By-value binding for a non-copy type where we copy into a
+    /// fresh LLVM alloca; this most accurately reflects the language
+    /// semantics (e.g. it properly handles overwrites of the matched
+    /// input), but potentially injects an unwanted copy.
+    TrByMoveIntoCopy(/* llbinding */ ValueRef),
+
+    /// Binding a non-copy type by reference under the hood; this is
+    /// a codegen optimization to avoid unnecessary memory traffic.
+    TrByMoveRef,
```

Comment:
Well, if we conflate `TrByCopy` and `TrByMoveIntoCopy`, then I think we end up back where we started (though perhaps with different names overall.

The only place where I think the handling of `TrByCopy` and `TrByMoveIntoCopy` differs in what kind of `Lvalue` I construct in `insert_lllocals`; everywhere else the two are collapsed.

(I personally found it nice to separate them just because I so strongly connect the word `Copy` with types that actually implement `Copy`...)